src/host/multiple_ledger.h

```diff
@@ -0,0 +1,815 @@
+            files.erase(f_);
+          }
+          else
+          {
+            f++;
+          }
+        }
+
+        // Continue writing at the end of last file only if that file is not
+        // complete
+        if (files.size() > 0 && !files.back()->is_complete())
+        {
+          require_new_file = false;
+        }
+        else
+        {
+          require_new_file = true;
+        }
+      }
+      else
+      {
+        if (!fs::create_directory(ledger_dir))
+        {
+          throw std::logic_error(fmt::format(
+            "Error: Could not create ledger directory: {}", ledger_dir));
+        }
+        require_new_file = true;
+      }
+    }
+
+    MultipleLedger(const MultipleLedger& that) = delete;
+
+    const std::vector<uint8_t> read_entry(size_t idx)
+    {
+      auto f = get_file_from_idx(idx);
+      if (f == nullptr)
+      {
+        return {};
+      }
+      return f->read_entry(idx);
+    }
+
+    const std::vector<uint8_t> read_framed_entries(size_t from, size_t to)
+    {
+      if ((from <= 0) || (to > last_idx) || (to < from))
+      {
+        return {};
+      }
+
+      std::vector<uint8_t> entries;
+      size_t idx = from;
+      while (idx <= to)
+      {
+        auto f_from = get_file_from_idx(idx);
+        if (f_from == nullptr)
+        {
+          return {};
+        }
+        auto to_ = std::min(f_from->get_last_idx(), to);
+        auto v = f_from->read_framed_entries(idx, to_);
+        if (v.size() == 0)
+        {
+          return entries;
+        }
+        entries.insert(
+          entries.end(),
+          std::make_move_iterator(v.begin()),
+          std::make_move_iterator(v.end()));
+        idx = to_ + 1;
+      }
+
+      return entries;
+    }
+
+    size_t write_entry(const uint8_t* data, size_t size, bool committable)
+    {
+      if (require_new_file)
+      {
+        files.push_back(std::make_shared<LedgerFile>(ledger_dir, last_idx + 1));
+        require_new_file = false;
+      }
+      auto f = get_latest_file();
+      last_idx = f->write_entry(data, size, committable);
+
+      LOG_DEBUG_FMT(
+        "Wrote entry at {} [committable: {}]", last_idx, committable);
+
+      if (committable && f->get_current_size() >= chunk_threshold)
+      {
+        f->complete();
+        require_new_file = true;
+        LOG_TRACE_FMT("New ledger chunk will start at {}", last_idx + 1);
+      }
+
+      return last_idx;
+    }
+
+    void truncate(size_t idx)
+    {
+      LOG_DEBUG_FMT("Ledger truncate: {}/{}", idx, last_idx);
+
+      if (idx >= last_idx || idx < compacted_idx)
+      {
+        return;
+      }
+
+      require_new_file = true;
+
+      auto f_from = get_it_contains_idx(idx + 1);
+      auto f_to = get_it_contains_idx(last_idx);
+      auto f_end = std::next(f_to);
+
+      for (auto it = f_from; it != f_end;)
+      {
+        // Truncate the first file to the truncation index while the more recent
+        // files are deleted entirely
```

Comment:
Is this right? Isn't it possible to truncate to the middle of a later file, if we don't get a commit for a long time?

Reply:
Unless I'm missing something, f_from is queried as "the first file that contains idx+1", so this does seem right.

Reply:
Ah cool, I'd missed this was only operating on a suffix range.